### PR TITLE
C# final code.

### PR DESCRIPTION
**Pull request by silverspoon19931120**
## Types of changes

_What types of changes does your code introduce to this project? Put an `x` in the boxes that apply._

- [ ] Bugfix.
- [x] New feature.
- [ ] Documentation update.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the pull request. If you are unsure about any of them, do not hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) documentation.
- [x] I have rebased, and my remote branch is now up-to-date with the `master` branch of the upstream.
- [x] Lint and unit tests pass locally with my changes.
- [x] I have added tests that prove my changes are effective.
- [ ] I have added the necessary documentation (if appropriate).

### CODE REVIEW

File: src/templates/csharp/TestBase.cs [hunk 1]
```diff
@@ -0,0 +1,1066 @@
...
+        }
+
+        public Uri GetAppiumServerUrl()
+        {
+            return new Uri(Config.AppiumServerUrl);
+        }
+
+        public long GetCurrentCommandId()
+        {
+            return this.proxy != null ? this.proxy.currentCommandId : 0;
```
Comment (chuong777) [inline src/templates/csharp/TestBase.cs:889]: `this.proxy` has not been initialized

File: src/templates/csharp/TestApp.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        [SetUp]
+        public void BeforeAll() {
+            Assert.AreNotEqual(
+                "Please update value for the KobitonApiKey constant first. See more at README.md file.",
+                "your_kobiton_api_key",
+                Config.KobitonApiKey
+            );
+        }
+
+        {{testCases}}
```
Comment (chuong777) [inline src/templates/csharp/TestApp.cs:27]: The generated project has empty test case body. It should call FindOnlineDevice, Setup, RunTest
image

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: src/templates/csharp/TestApp.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                UpdateSettings();
+                SwitchToNativeContext();
+                SetImplicitWaitInMiliSecond(Config.ImplicitWaitInMs);
+                {{testScript}}
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e.StackTrace);
+                SaveDebugResource();
+                throw e;
```
Comment (chuong777) [inline src/templates/csharp/TestApp.cs:41]:
```suggestion
```

We need to call `Cleanup` to end session whether an exception occurs or not

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: src/templates/csharp/TestBase.cs [hunk 4]
```diff
@@ -0,0 +1,1066 @@
...
+                httpClient.SendAsync(request).Wait();
+            } else
+            { 
+                for (int index = 0; index < keys.Count(); index++)
+                {
+                    int codePoint = Char.ConvertToUtf32(keys, index);
+                    GetAndroidDriver().PressKeyCode(codePoint);
+                    sleep(SleepAfterAction);
+                }
+            }
```
Comment (chuong777) [inline src/templates/csharp/TestBase.cs:612]: SendKeys is not work as expected on Android. It send unexpected key code that make the application go to background.

Java implementation:
image

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: __tests__/resource/csharp-nunit-input.json [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+              "value": "//android.widget.EditText[@resource-id='com.android.chrome:id/url_bar']"
+            },
+            {
+              "type": "xpath",
+              "value": "//android.widget.EditText[@text='edition.cnn.com']"
+            }
+          ]
+        }
+      ],
+      "actionJson": "{\"command\":\"touchOnElement\",\"x\":\"0.387\",\"y\":\"0.430\",\"value\":\"beyoung.in\"}",
```
Comment (chuong777) [inline __tests__/resource/csharp-nunit-input.json:56]: on an `touchOnElement ` we cannot have "value" field, please keep it as a separated test step with command "sendKey", sample: https://github.com/kobiton/appium-script-generator/blob/master/__tests__/resource/java-junit-input.json#L63

File: src/services/csharp.js [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+            const elementVarName = `element${rawLocatorVarName}`
+            // eslint-disable-next-line max-len
+            lines.push(new Line(`AppiumWebElement ${elementVarName} = FindElementBy(${findingElementTimeout}, ${locatorVarName});`))
+            // eslint-disable-next-line max-len
+            lines.push(new Line(`TouchOnElementByType(${elementVarName}, ${x}, ${y});`))
+            //
+            try {
+              const {value} = action
+              if (value != undefined) {
+                lines.push(new Line(`${elementVarName}.SendKeys(${this._getString(value)});`))
```
Comment (chuong777) [inline src/services/csharp.js:318]: on "touchOnElement" step we cannot have "value" field for sendKeys command, it should be a separated step

**Review by chuong777 — CHANGES_REQUESTED** [review]

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: src/services/csharp.js [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+          const {x1, y1, x2, y2, duration} = action
+          if (context === CONTEXTS.NATIVE) {
+            /* eslint-disable */
+            !isOnKeyboard && lines.push(new Line('HideKeyboard();'))
+
+            const elementVarName = `element${rawLocatorVarName}`
+            lines.push(new Line(`AppiumWebElement ${elementVarName} = FindElementBy(${findingElementTimeout}, ${locatorVarName});`))
+
+            const rectVarName = `rect${rawLocatorVarName}`
+            lines.push(new Line(`Rectangle ${rectVarName} = GetWebElementRect(${elementVarName});`))
```
Comment (chuong777) [inline src/services/csharp.js:340]: When context is NATIVE we will use other method (please check java code)

**Review by chuong777 — CHANGES_REQUESTED** [review]

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: __tests__/resource/csharp-nunit-input.json [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+      "id": "17001",
+      "context": "NATIVE",
+      "actionJson": "{\"command\":\"sendKeys\",\"value\":\"cnn.com\"}",
+      "findingElementTimeout": 4014,
+      "isOnKeyboard": false
+    },
+    {
+      "id": "17002",
+      "context": "NATIVE",
+      "actionJson": "{\"command\":\"press\",\"value\":\"Enter\"}",
```
Comment (chuong777) [inline __tests__/resource/csharp-nunit-input.json:70]:
```suggestion
      "actionJson": "{\"command\":\"press\",\"value\":\"ENTER\"}",
```
Please copy `testSteps` data from java file then run unit test again and check if project can be build

File: src/handlers/generate-script.js [hunk 9]
```diff
@@ -42,7 +42,7 @@ const generateHandler = async ({request}, callback) => {
     }
 
     // Clean up cached files
-    await removeDir(workingDir)
+    //await removeDir(workingDir)
```
Comment (chuong777) [inline src/handlers/generate-script.js:45]:
```suggestion
    await removeDir(workingDir)
```

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: src/templates/csharp/TestApp.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+{
+    public class Tests : TestBase
+    {
+        [SetUp]
+        public void BeforeAll() {
+            Assert.AreNotEqual(
+                "Please update value for the KobitonApiKey constant first. See more at README.md file.",
+                "your_kobiton_api_key",
+                Config.KobitonApiKey
+            );
```
Comment (chuong777) [inline src/templates/csharp/TestApp.cs:15]: This assertion is not work

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: src/templates/csharp/TestApp.cs (hunk 10 above)
Comment (chuong777) [inline src/templates/csharp/TestApp.cs:15]: This assertion is not working (wrong order of parameters). Please use `Assert.That `instead
```suggestion
            Assert.That(
                Config.KobitonApiKey, Is.Not.EqualTo("your_kobiton_api_key"),
                "Please update value for the KobitonApiKey constant first. See more at README.md file."
            );
```

**Review by chuong777 — CHANGES_REQUESTED** [review]

File: __tests__/integration/generate-script.test.js [hunk 11]
```diff
@@ -8,11 +8,13 @@ import generateScriptHandler from '../../src/handlers/generate-script'
...
-  __dirname, '../../src/appium-script-schema/generate-script.proto')
+    __dirname, '../../src/appium-script-schema/generate-script.proto')
 const JAVA_JUNIT_INPUT_FILE = path.resolve(
-  __dirname, '../resource/java-junit-input.json')
+    __dirname, '../resource/java-junit-input.json')
+  const CSHARP_NUNIT_INPUT_FILE = path.resolve(
+    __dirname, '../resource/csharp-nunit-input.json')
 const NODEJS_MOCHA_INPUT_FILE = path.resolve(
-  __dirname, '../resource/nodejs-mocha-input.json')
+    __dirname, '../resource/nodejs-mocha-input.json')
```
Comment (ledinhphuong) [inline __tests__/integration/generate-script.test.js:17]: Please correct the indentation

File: src/templates/csharp/TestBase.cs [hunk 12]
```diff
@@ -0,0 +1,1100 @@
...
+            );
+
+            return rect;
+        }
+
+        public Rectangle CalculateNativeRect(Rectangle webElementRect)
+        {
+            AppiumWebElement appiumWebElement = FindWebview();
+            Rectangle webviewRect = new Rectangle(
+                appiumWebElement).Location.X,
```
Comment (ledinhphuong) [inline src/templates/csharp/TestBase.cs:249]:
```suggestion
                appiumWebElement.Location.X,
```

**Review by ledinhphuong — CHANGES_REQUESTED** [review]

File: src/templates/csharp/TestBase.cs [hunk 13]
```diff
@@ -0,0 +1,1100 @@
...
+        }
+
+        public async void Cleanup()
+        {
+            driver.Quit();
+
+            if(proxy != null)
+            {
+                proxy.StopProxy();
+            }
```
Comment (ledinhphuong) [inline src/templates/csharp/TestBase.cs:71]:
```suggestion
            if (driver != null)
            {
                driver.Quit();
            }

            if (proxy != null)
            {
                proxy.StopProxy();
            }
```

**Review by ledinhphuong — CHANGES_REQUESTED** [review]

File: src/templates/csharp/Config.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OpenQA.Selenium.Appium;
+
+namespace AppiumTest
+{
+    public class Config
+    {
+        public enum DeviceSourceEnums { KOBITON, SAUCE_LABS }
```
Comment (chuong777) [inline src/templates/csharp/Config.cs:13]: Use PascalCasing for enum

File: src/templates/csharp/TestBase.cs [hunk 15]
```diff
@@ -0,0 +1,1103 @@
...
+using Newtonsoft.Json.Linq;
+using System.Text.RegularExpressions;
+using System.Net;
+using OpenQA.Selenium.Interactions;
+
+namespace AppiumTest
+{
+    public class TestBase
+    {
+        public enum PressTypes { HOME, BACK, POWER, APP_SWITCH, ENTER, DELETE }
```
Comment (chuong777) [inline src/templates/csharp/TestBase.cs:23]: Use PascalCasing for enum

**Review by chuong777 — CHANGES_REQUESTED** [review]

**Review by ledinhphuong — CHANGES_REQUESTED** [review]

**Review by chuong777 — APPROVED** [review]

**Review by ledinhphuong — APPROVED** [review]

**Comment by silverspoon19931120** [comment]
HI Phoung?
How about my c# script works?
and can we continue next ruby script?
Frank said to me we can start in 22/ January.

**Comment by silverspoon19931120** [comment]
HI, Phoung, there?.

**Pull request merged**